compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -846,6 +846,132 @@ fn check_impl_item(tcx: TyCtxt<'_>, impl_item: &hir::ImplItem<'_>) {
+    if tcx.features().adt_const_params {
+        if let Some(non_structural_match_ty) =
+            traits::search_for_adt_const_param_violation(param_span, tcx, ty)
+        {
+            // We use the same error code in both branches, because this is really the same
+            // issue: we just special-case the message for type parameters to make it
+            // clearer.
+            match non_structural_match_ty.kind() {
+                ty::Param(_) => {
+                    // Const parameters may not have type parameters as their types,
+                    // because we cannot be sure that the type parameter derives `PartialEq`
+                    // and `Eq` (just implementing them is not enough for `structural_match`).
+                    struct_span_err!(
+                        tcx.sess,
+                        hir_ty.span,
+                        E0741,
+                        "`{ty}` is not guaranteed to `#[derive(PartialEq, Eq)]`, so may not be \
+                        used as the type of a const parameter",
+                    )
+                    .span_label(
+                        hir_ty.span,
+                        format!("`{ty}` may not derive both `PartialEq` and `Eq`"),
+                    )
+                    .note(
+                        "it is not currently possible to use a type parameter as the type of a \
+                        const parameter",
+                    )
+                    .emit();
+                }
+                ty::Float(_) => {
+                    struct_span_err!(
+                        tcx.sess,
+                        hir_ty.span,
+                        E0741,
+                        "`{ty}` is forbidden as the type of a const generic parameter",
+                    )
+                    .note("floats do not derive `Eq` or `Ord`, which are required for const parameters")
+                    .emit();
+                }
+                ty::FnPtr(_) => {
+                    struct_span_err!(
+                        tcx.sess,
+                        hir_ty.span,
+                        E0741,
+                        "using function pointers as const generic parameters is forbidden",
+                    )
+                    .emit();
+                }
+                ty::RawPtr(_) => {
+                    struct_span_err!(
+                        tcx.sess,
+                        hir_ty.span,
+                        E0741,
+                        "using raw pointers as const generic parameters is forbidden",
+                    )
+                    .emit();
+                }
+                ty::Projection(ty::ProjectionTy { substs, item_def_id }) => {
+                    let binder_ty = tcx.bound_type_of(*item_def_id);
+                    let ty = binder_ty.subst(tcx, substs);
+                    check_param_wf_ty(tcx, ty, hir_ty, param_span);
+                }
```

Comment:
this seems like the wrong place to me. `search_for_adt_const_param_violation` should probably be normalizing at every step of the way. cc @rust-lang/project-const-generics

Reply:
I'll add a commit where I do it directly into `search_for_adt_const_param_violation` visitor and check if it works as expected.

Reply:
So after discussion with @oli-obk, until we need it, we won't put this normalization into `search_for_adt_const_param_violation`.